packages/test/src/workflows/signals-are-always-processed.ts

```diff
@@ -0,0 +1,19 @@
+import { proxySinks, Sinks, setHandler, defineSignal } from '@temporalio/workflow';
+
+export const incrementSignal = defineSignal('increment');
+
+export interface SignalProcessTestSinks extends Sinks {
+  controller: {
+    sendSignal(): void;
+  };
+}
+
+const { controller } = proxySinks<SignalProcessTestSinks>();
+
+export async function signalsAreAlwaysProcessed(): Promise<void> {
+  let counter = 0;
+  setHandler(incrementSignal, () => void ++counter);
+  if (counter === 0) {
```

Comment:
I do not understand the workflow here (sorry a bit unfamiliar w/ the TS SDK). How will this not always be zero if the counter is not incremented until after this?

Reply:
The counter is incremented in the signal handler.
Signals are buffered and delivered immediately on handler registration.

Reply:
Hrmm, reading the code one might expect without awaits anywhere that it would immediately run and return. In normal JS (browser, node, etc) I don't think one would expect the callback in `setHandler` to run before the next statement.

Reply:
We have to deliver the signal otherwise it'd be lost.

Reply:
I see. I think this may be a confusion in the API. Arguably signals should not be a handler that you don't know when it will happen, but rather something that can be awaited on at a point in the workflow where you care. One might expect to be able to `await signal.receive()`. As it is now I would probably add this helper (untested/pseudo) for myself:

```typescript
waitSignal<T>(signal: SignalDefinition<T>): Promise<T> {
  return new Promise(res => setHandler(signal, res))
}
```

The other SDKs let you choose exactly where to wait for signals. For example, take a workflow example in Go from this morning at https://community.temporal.io/t/signalling-workflows/3370/8. That user wants to receive a signal in a loop, waiting to execute the the activity until after each signal.

Reply:
To confirm, maybe it's Go that's weird that lets you choose each time to receive a signal value instead of having a signal forced upon you. I see Java also operates on this signal handler mechanism.